nixos/tests/matrix/matrix-alertmanager.nix

```diff
@@ -0,0 +1,190 @@
+    csr = runWithOpenSSL "matrix.csr" ''
+      openssl req \
+         -new -key ${key} \
+         -out $out -subj "/CN=localhost" \
+    '';
+    cert = runWithOpenSSL "matrix_cert.pem" ''
+      openssl x509 \
+        -req -in ${csr} \
+        -CA ${ca_pem} -CAkey ${ca_key} \
+        -CAcreateserial -out $out \
+        -days 365
+    '';
+  in
+  {
+    name = "matrix-alertmanager";
+    meta.maintainers = with pkgs.lib.maintainers; [ erethon ];
+
+    nodes = {
+      homeserver =
+        { pkgs, ... }:
+        {
+          services.matrix-synapse = {
+            enable = true;
+            settings = {
+              database.name = "sqlite3";
+              tls_certificate_path = "${cert}";
+              tls_private_key_path = "${key}";
+              enable_registration = true;
+              enable_registration_without_verification = true;
+              registration_shared_secret = "supersecret-registration";
+
+              listeners = [
+                {
+                  # The default but tls=false
+                  bind_addresses = [
+                    "0.0.0.0"
+                  ];
+                  port = 8448;
+                  resources = [
+                    {
+                      compress = true;
+                      names = [ "client" ];
+                    }
+                    {
+                      compress = false;
+                      names = [ "federation" ];
+                    }
+                  ];
+                  tls = false;
+                  type = "http";
+                  x_forwarded = false;
+                }
+              ];
+            };
+          };
+
+          networking.firewall.allowedTCPPorts = [ 8448 ];
+
+          environment.systemPackages = [
+            (pkgs.writeShellScriptBin "register_alertmanager_user" ''
+              exec ${pkgs.matrix-synapse}/bin/register_new_matrix_user \
+                -u alertmanager \
+                -p alertmanager-password \
+                --admin \
+                --shared-secret supersecret-registration \
+                http://localhost:8448
+            '')
+            # This is needed to solve a chicken and egg
+            # problem. Matrix-alertmanager expects a token for authentication,
+            # but a token is create after the user has been registered. This
+            # changes the token in the database to match the one specified in
+            # the service settings.
+            (pkgs.writers.writePython3Bin "hardcode_matrix_values"
+              {
+                libraries = with pkgs.python3Packages; [
+                  sqlite-utils
+                ];
+              }
+              ''
+                import sqlite3
+                con = sqlite3.connect("/var/lib/matrix-synapse/homeserver.db")
+                cur = con.cursor()
+                cur.execute("update access_tokens set token='faketoken' where id = 2")
```

Comment:
nit: Any chance we could write this query using the username instead of a hardcoded id? (I don't know the schema of this database at all, feel free to discard if this is too complicated)

Reply:
Rewrote this in a less ambigious way now that there's only a single token for this user.